devimint/src/external.rs

```diff
@@ -240,11 +242,73 @@ impl Bitcoind {
     }
 
-    pub async fn get_raw_transaction(&self, txid: bitcoin30::Txid) -> Result<String> {
+    /// Get a transaction by its txid. Checks the mempool and all blocks.
+    pub async fn get_transaction(&self, txid: bitcoin30::Txid) -> Result<Option<String>> {
+        // Check the mempool.
+        match self.get_raw_transaction(txid, None).await {
+            // The RPC succeeded, and the transaction was not found in the mempool. Continue to
+            // check blocks.
+            Ok(None) => {}
+            // The RPC failed, or the transaction was found in the mempool. Return the result.
+            other => return other,
+        };
+
+        let block_height = self.get_block_count().await? - 1;
+
+        // Check each block for the tx, starting at the chain tip.
+        // Buffer the requests to avoid spamming bitcoind.
+        // We're doing this after checking the mempool since the tx should
+        // usually be in the mempool, and we don't want to needlessly hit
+        // the bitcoind with block requests.
+        let mut buffered_tx_stream = futures::stream::iter((0..block_height).rev())
+            .map(|height| async move {
+                let block_hash = self.get_block_hash(height).await?;
+                self.get_raw_transaction(txid, Some(block_hash)).await
+            })
+            .buffered(32);
+
+        while let Some(tx_or) = buffered_tx_stream.next().await {
+            match tx_or {
+                // The RPC succeeded, and the transaction was not found in the block. Continue to
+                // the next block.
+                Ok(None) => continue,
+                // The RPC failed, or the transaction was found in the block. Return the result.
+                other => return other,
+            };
+        }
+
+        Err(anyhow!("Transaction not found"))
+    }
+
+    async fn get_raw_transaction(
+        &self,
+        txid: bitcoin30::Txid,
+        block_hash: Option<BlockHash>,
+    ) -> Result<Option<String>> {
         let client = self.client.clone();
-        let tx = spawn_blocking(move || client.get_raw_transaction(&txid, None)).await??;
+        let tx_or =
+            spawn_blocking(move || client.get_raw_transaction(&txid, block_hash.as_ref())).await?;
+
+        // `getrawtransaction` returns a JSON-RPC error with code -5 if the command
+        // reaches bitcoind but is not found. See here:
+        // https://github.com/bitcoin/bitcoin/blob/25dacae9c7feb31308271e2fd5a127c1fc230c2f/src/rpc/rawtransaction.cpp#L360-L376
+        // https://github.com/bitcoin/bitcoin/blob/25dacae9c7feb31308271e2fd5a127c1fc230c2f/src/rpc/protocol.h#L42
+        let tx = match tx_or {
+            Ok(tx) => tx,
+            Err(err) => match err {
+                bitcoincore_rpc::Error::JsonRpc(bitcoincore_rpc::jsonrpc::Error::Rpc(
+                    RpcError { code: -5, .. },
+                )) => return Ok(None),
+                _ => return Err(err.into()),
```

Comment:
I wonder if we should log here. Some callers ignore the error case, which should never happen imo. It's probably good to not panic to avoid flakes, but if there is consistent failure reported from `bitcoind` we'd want to see it in the test logs.

Reply:
I fixed this by adding `poll_get_transaction()` so all callers exit on the error case while continuing to poll on the `Ok(None)` case. I don't think we need to log here, since `poll_get_transaction()` calls `.context("getrawtransaction")` on the error case. That should be enough to see what's happening, right?